src/sinks/util/sink.rs

```diff
@@ -464,27 +469,33 @@ where
         let partition = item.partition();
 
         if let Some(batch) = self.partitions.get_mut(&partition) {
-            if batch.len() >= self.settings.size {
+            if batch.was_full() {
                 trace!("Batch full; driving service to completion.");
                 self.poll_complete()?;
 
                 if let Some(batch) = self.partitions.get_mut(&partition) {
-                    if batch.len() >= self.settings.size {
+                    if !batch.is_empty() {
                         debug!(
                             message = "Buffer full; applying back pressure.",
-                            max_size = %self.settings.size,
+                            max_bytes = %self.settings.bytes,
+                            max_events = %self.settings.events,
                             rate_limit_secs = 10
                         );
                         return Ok(AsyncSink::NotReady(item));
                     } else {
-                        batch.push(item);
-                        return Ok(AsyncSink::Ready);
+                        return match batch.push(item) {
+                            PushResult::Ok => Ok(AsyncSink::Ready),
+                            // FIXME is this the right behavior here and below?
+                            PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
+                        };
                     }
                 }
             } else {
                 trace!("adding event to batch.");
-                batch.push(item);
-                return Ok(AsyncSink::Ready);
+                return match batch.push(item) {
+                    PushResult::Ok => Ok(AsyncSink::Ready),
+                    PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
```

Comment:
Here I'm less sure. We should not return `NotReady` without trying to drive things forward via `poll_complete`, and it doesn't seem like we've done that in this branch.

Reply:
Is that just a matter of adding a `self.poll_complete()?` here, or something more?

Reply:
The rule is that we shouldn't return `PollReady` unless something below us has returned it. The idea is that goes all the way down to IO or timer operations that register our task to be woken up at the appropriate future time. So we should be doing as much work as possible before we're blocked by `NotReady`.

In this case, that specifically means we should call `poll_complete` and then either `push` again if it returns `Ready` (since that should have cleared us room) or return `NotReady` if it returns the same to us.

It seems to me that some reorganization of this method might be useful. We have a few places with code that seems unreachable and doing this properly is probably going to add some duplication.

Reply:
I tried to do it properly, but couldn't figure any good way without either some duplication, or introducing some enums just for use within this method.